cpp/src/rolling/grouped_rolling.cu

```diff
@@ -269,6 +270,69 @@ __device__ T subtract_safe(T const& value, T const& delta)
            : cuda::std::numeric_limits<T>::min();
 }
 
+/**
+ * @brief For a specified idx, find the lowest value of the (sorted) orderby column that
+ * participates in a range-window query.
+ */
+template <typename ElementT, typename ElementIter>
+__device__ ElementT compute_lowest_in_window(ElementIter orderby_iter,
+                                             size_type idx,
+                                             ElementT delta)
+{
+  return subtract_safe(orderby_iter[idx], delta);
+}
+
+/**
+ * @brief For a specified idx, find the highest value of the (sorted) orderby column that
+ * participates in a range-window query.
+ */
+template <typename ElementT, typename ElementIter>
+__device__ ElementT compute_highest_in_window(ElementIter orderby_iter,
+                                              size_type idx,
+                                              ElementT delta)
+{
+  return add_safe(orderby_iter[idx], delta);
+}
+
+/**
+ * Accessor for values in an order-by column, on the device.
+ */
+template <typename T>
+struct device_value_accessor {
+  column_device_view const col;  ///< column view of column in device
+
+  /**
+   * @brief constructor
+   *
+   * @param[in] col_ column device view of cudf column
+   */
+  __device__ device_value_accessor(column_device_view const& col_) : col{col_}
+  {
+    cudf_assert(type_id_matches_device_storage_type<T>(col.type().id()) &&
+                "the data type mismatch");
+  }
+
+  /**
+   * @brief Returns the value of element at index `i`
+   * @param[in] i index of element
+   * @return value of element at index `i`
+   */
+  __device__ T operator()(cudf::size_type i) const { return col.element<T>(i); }
+};
+
+template <typename T>
+using const_device_iterator =
+  thrust::transform_iterator<device_value_accessor<T>, thrust::counting_iterator<size_type>>;
+
+/// This is a stand-in for the `cudf::column_device_view::begin<T>()`, which is `__host__` only.
+/// For range window functions, one might need to iterate over the `oby` column, per row.
+template <typename T, CUDF_ENABLE_IF(cudf::column_device_view::has_element_accessor<T>())>
+[[nodiscard]] __device__ const_device_iterator<T> begin(cudf::column_device_view const& col)
```

Comment:
Sorry I don't see this being used anywhere?

Reply:
[Here](https://github.com/rapidsai/cudf/pull/13095/files#diff-43139648ae3ec9db09ede1dde96e6e5c88d67a32320a07c7a1dfa267ff918291R410) is one example. It is used in the preceding/following calculation instead of using `d_orderby = orderby_column.data<T>()`, so as to ultimately use `.element<T>()`.